coordinator/controllers/node_controller.go

```diff
@@ -62,316 +65,307 @@ type NodeController interface {
+		n.maybeInitHealthClient()
+		ticker := time.NewTicker(healthCheckProbeInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-n.ctx.Done():
+				return nil
+			case <-ticker.C:
+				pingCtx, pingCancel := context.WithTimeout(n.ctx, healthCheckProbeTimeout)
+				response, err := n.healthClient.Check(pingCtx, &grpc_health_v1.HealthCheckRequest{Service: ""})
+				pingCancel()
+				if err := n.healthCheckHandler(response, err); err != nil {
+					n.Warn("Node stopped responding to ping")
+					return err
+				}
+			}
 		}
+	}, n.healthCheckBackoff, func(err error, duration time.Duration) {
+		n.Warn(
+			"Failed to check storage node heath by ping-pong",
+			slog.Duration("retry-after", duration),
+			slog.Any("error", err),
+		)
+		n.becomeUnavailable()
 	})
 }
 
-func (n *nodeController) healthCheckLoop(ctx context.Context, health grpc_health_v1.HealthClient) {
-	ticker := time.NewTicker(healthCheckProbeInterval)
-
-	for {
-		select {
-		case <-ticker.C:
-			pingCtx, pingCancel := context.WithTimeout(ctx, healthCheckProbeTimeout)
+func (n *nodeController) healthWatchWithRetries() {
+	defer n.Done()
+	_ = backoff.RetryNotify(func() error {
+		n.Debug("Start new health watch cycle")
+		n.maybeInitHealthClient()
 
-			res, err := health.Check(pingCtx, &grpc_health_v1.HealthCheckRequest{Service: ""})
-			pingCancel()
-			if err2 := n.processHealthCheckResponse(res, err); err2 != nil {
-				n.log.Warn("Node stopped responding to ping")
-				return
+		watchStream, err := n.healthClient.Watch(n.ctx, &grpc_health_v1.HealthCheckRequest{Service: ""})
+		if err != nil {
+			return err
+		}
+		for {
+			select {
+			case <-n.ctx.Done():
+				return nil
+			default:
+				if err := n.healthCheckHandler(watchStream.Recv()); err != nil {
+					return err
+				}
 			}
-
-		case <-ctx.Done():
-			return
 		}
-	}
+	}, n.healthWatchBackoff, func(err error, duration time.Duration) {
+		n.Warn("Failed to check storage node heath by watch",
+			slog.Any("error", err),
+			slog.Duration("retry-after", duration),
+		)
+		n.becomeUnavailable()
+	})
 }
 
-func (n *nodeController) healthCheck(backoff backoff.BackOff) error {
-	n.log.Debug("Start new health check cycle")
-	health, closer, err := n.rpc.GetHealthClient(n.server)
-	if err != nil {
-		n.log.Debug("Failed to get health check client", slog.Any("error", err))
-		return err
+func (n *nodeController) becomeUnavailable() {
+	currentStatus := n.Status()
+	if currentStatus != Running && currentStatus != Draining {
+		return
 	}
-
-	defer closer.Close()
-
-	ctx, cancel := context.WithCancel(n.ctx)
-	defer cancel()
-
-	go process.DoWithLabels(
-		ctx,
-		map[string]string{
-			"oxia":   "node-controller-health-check-ping",
-			"server": n.server.GetIdentifier(),
-		},
-		func() {
-			defer cancel()
-			n.healthCheckLoop(ctx, health)
-		},
-	)
-
-	watch, err := health.Watch(ctx, &grpc_health_v1.HealthCheckRequest{Service: ""})
-	if err != nil {
-		return err
+	n.statusLock.Lock()
```

Comment:
There is already a lock operation in n.Status. I don't think we need double lock here.